projects/extension/sql/idempotent/904-semantic-catalog-search.sql

```diff
@@ -0,0 +1,241 @@
+--FEATURE-FLAG: text_to_sql
+
+-------------------------------------------------------------------------------
+-- _semantic_catalog_embed
+create or replace function ai._semantic_catalog_embed
+( catalog_id pg_catalog.int4
+, prompt pg_catalog.text
+) returns @extschema:vector@.vector
+as $func$
+declare
+    _vectorizer_id pg_catalog.int4;
+    _config pg_catalog.jsonb;
+    _emb @extschema:vector@.vector;
+begin
+    select x.obj_vectorizer_id -- TODO: assumes the embedding settings are the same for obj and sql
+    into strict _vectorizer_id
+    from ai.semantic_catalog x
+    where x.id operator(pg_catalog.=) catalog_id
+    ;
+
+    select v.config operator(pg_catalog.->) 'embedding'
+    into strict _config
+    from ai.vectorizer v
+    where v.id operator(pg_catalog.=) _vectorizer_id
+    ;
+
+    case _config operator(pg_catalog.->>) 'implementation'
+        when 'openai' then
+            _emb = ai.openai_embed
+            ( _config operator(pg_catalog.->>) 'model'
+            , prompt
+            , api_key_name=>(_config operator(pg_catalog.->>) 'api_key_name')
+            , dimensions=>(_config operator(pg_catalog.->>) 'dimensions')::pg_catalog.int4
+            , openai_user=>(_config operator(pg_catalog.->>) 'user')
+            );
+        when 'ollama' then
+            _emb = ai.ollama_embed
+            ( _config operator(pg_catalog.->>) 'model'
+            , prompt
+            , host=>(_config operator(pg_catalog.->>) 'base_url')
+            , keep_alive=>(_config operator(pg_catalog.->>) 'keep_alive')
+            , embedding_options=>(_config operator(pg_catalog.->) 'options')
+            );
+        when 'voyageai' then
+            _emb = ai.voyageai_embed
+            ( _config operator(pg_catalog.->>) 'model'
+            , prompt
+            , input_type=>'query'
+            , api_key_name=>(_config operator(pg_catalog.->>) 'api_key_name')
+            );
+        else
+            raise exception 'unsupported embedding implementation';
+    end case;
+
+    return _emb;
+end
+$func$ language plpgsql stable security invoker
+set search_path to pg_catalog, pg_temp
+;
+
+-------------------------------------------------------------------------------
+-- find_relevant_sql
+create or replace function ai._find_relevant_sql
+( catalog_id pg_catalog.int4
+, embedding @extschema:vector@.vector
+, "limit" pg_catalog.int8 default 5
```

Comment:
Why provide limit here? And can this function be plain sql? I assume (naively) that

1. If the function would be a plain sql function returning a table, then the called can use `LIMIT x` instead of having to pass the limit in as a param
2. That the planner will be able to "see through" the plain sql function so this would not be horribly inefficient.

Reply:
>can this function be plain sql?

No. We are querying one of the potentially multiple views. `from ai.semantic_catalog_sql_%s x` So, the SQL has to be generated dynamically. We are supporting multiple vectorizers on the semantic catalog.

>Why provide limit here?

Since this cannot be a plain SQL function, we need to put the limit in the generated query for performance. Otherwise, it will return ALL the rows in the table. Yeah, like you said, I don't think there's any way for the planner to "push down" that limit.